File: rock/ts-sdk/src/job/executor.ts
/**
 * JobExecutor — orchestrates the full execution of Trials produced by an Operator.
 *
 * Flow:
 *     submit(operator, config)  — apply operator to get TrialList, start all sandboxes
 *                                  in parallel, return JobClient (list of TrialClient)
 *     wait(job_client)          — wait for all trials, collect results, return list[TrialResult[]]
 *     run(operator, config)     — submit + wait
 *
 * Matches Python rock.sdk.job.executor.
 */

import type { AbstractTrial } from './trial/abstract';
import type { Operator } from './operator';
import { USER_DEFINED_LOGS } from '../bench/constants';
import { Sandbox } from '../sandbox/client';
import { ExceptionInfoSchema, type TrialResult } from './result';
import type { Observation, ReadFileResponse } from '../types/responses';
import type { CreateBashSessionRequest, WriteFileRequest, ReadFileRequest } from '../types/requests';
import { shellQuote } from '../utils/shell';

// ---------------------------------------------------------------------------
// TrialClient / JobClient
// ---------------------------------------------------------------------------

/** Handle for a single running trial. */
export interface TrialClient {
  sandbox: JobSandbox;
  session: string;
  pid: number;
  trial: AbstractTrial;
}

/** Handle returned by JobExecutor.submit(). Holds multiple TrialClients. */
export interface JobClient {
  trials: TrialClient[];
}

export interface JobSandbox {
  start(): Promise<void>;
  getNamespace(): string | null;
  getExperimentId(): string | null;
  createSession(request: CreateBashSessionRequest): Promise<unknown>;
  writeFile(request: WriteFileRequest): Promise<{ success: boolean; message?: string }>;
  readFile(request: ReadFileRequest): Promise<ReadFileResponse>;
  startNohupProcess(
    cmd: string,
    tmpFile: string,
    session: string
  ): Promise<{ pid: number | null; errorResponse: Observation | null }>;
  waitForProcessCompletion(
    pid: number,
    session: string,
    waitTimeout: number,
    waitInterval: number
  ): Promise<{ success: boolean; message: string }>;
  handleNohupOutput(
    tmpFile: string,
    session: string,
    success: boolean,
    message: string,
    ignoreOutput: boolean,
    responseLimitedBytesInNohup: number | null
  ): Promise<Observation>;
  arun?(cmd: string, options?: { session?: string }): Promise<unknown>;
}

export type SandboxFactory = (config: Record<string, unknown>) => JobSandbox;

// ---------------------------------------------------------------------------
// JobExecutor
// ---------------------------------------------------------------------------

export class JobExecutor {
  constructor(
    private readonly sandboxFactory: SandboxFactory = (config) => new Sandbox(config)
  ) {}

  /**
   * Full lifecycle: submit + wait.
   */
  async run(
    operator: Operator,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: Record<string, any>
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any[]> {
    const jobClient = await this.submit(operator, config);
    return this.wait(jobClient);
  }

  /**
   * Operator generates TrialList, start all sandboxes in parallel.
   */
  async submit(
    operator: Operator,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: Record<string, any>
  ): Promise<JobClient> {
    const trialList = operator.apply(config);
    if (trialList.length === 0) {
      return { trials: [] };
    }
    const trials = await Promise.all(trialList.map((trial) => this._doSubmit(trial)));
    return { trials };
  }

  /**
   * Wait for all trials, collect results in parallel.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async wait(jobClient: JobClient): Promise<any[]> {
    if (jobClient.trials.length === 0) {
      return [];
    }
    return Promise.all(jobClient.trials.map((tc) => this._doWait(tc)));
  }

  // ------------------------------------------------------------------
  // Private
  // ------------------------------------------------------------------

  private static jobTmpPrefix(config: Record<string, unknown>): string {
    return `${USER_DEFINED_LOGS}/rock_job_${config['job_name'] ?? 'default'}`;
  }

  private static jobExitPath(config: Record<string, unknown>): string {
    return `${JobExecutor.jobTmpPrefix(config)}.exit`;
  }

  private static buildRunScriptCommand(scriptPath: string, exitPath: string): string {
    const inner = [
      `bash ${shellQuote(scriptPath)}`,
      'rc=$?',
      `echo "$rc" > ${shellQuote(exitPath)}`,
      'exit "$rc"',
    ].join('; ');
    return `bash -c ${shellQuote(inner)}`;
  }

  private async _doSubmit(trial: AbstractTrial): Promise<TrialClient> {
    const config = trial.config;
    const sandbox = this.sandboxFactory((config['environment'] ?? {}) as Record<string, unknown>);

    await sandbox.start();
    await trial.onSandboxReady(sandbox);
    await trial.setup(sandbox as Sandbox);

    const session = `rock-job-${config['job_name'] ?? 'default'}`;
    const env = JobExecutor.buildSessionEnv(config);
    await sandbox.createSession({ session, startupSource: [], envEnable: true, env: env ?? undefined });

    const scriptPath = `${JobExecutor.jobTmpPrefix(config)}.sh`;
    const writeResult = await sandbox.writeFile({ content: trial.build(), path: scriptPath });
    if (!writeResult.success) {
      throw new Error(`Failed to write job script ${scriptPath}: ${writeResult.message ?? ''}`);
    }

    const tmpFile = `${JobExecutor.jobTmpPrefix(config)}.out`;
    const exitPath = JobExecutor.jobExitPath(config);
    const { pid, errorResponse } = await sandbox.startNohupProcess(
      JobExecutor.buildRunScriptCommand(scriptPath, exitPath),
      tmpFile,
      session
    );
    if (errorResponse) {
      throw new Error(`Failed to start trial: ${errorResponse.output || errorResponse.failureReason}`);
    }
    if (!pid) {
      throw new Error('Failed to start trial: nohup did not return a PID');
    }

    return { sandbox, session, pid, trial };
  }

  private async _doWait(client: TrialClient): Promise<TrialResult | TrialResult[]> {
    const config = client.trial.config;
    const { success, message } = await client.sandbox.waitForProcessCompletion(
      client.pid,
      client.session,
      config['timeout'] ?? 7200,
      30
    );
    const obs = await client.sandbox.handleNohupOutput(
      `${JobExecutor.jobTmpPrefix(config)}.out`,
      client.session,
      success,
      message,
      false,
      null
    );
    const exitCode = await this.readScriptExitCode(client, obs, success);
    const result = await client.trial.collect(client.sandbox as Sandbox, obs.output ?? '', exitCode);
    const results = Array.isArray(result) ? result : [result];

    for (const r of results) {
      if (!r.raw_output) {
        r.raw_output = obs.output ?? '';
      }
      if (r.exit_code === 0 && exitCode !== 0) {
        r.exit_code = exitCode;
      }
      if (!success && r.exception_info === null) {
        r.exception_info = ExceptionInfoSchema.parse({
          exception_type: 'ProcessTimeout',
          exception_message: message || 'process did not complete successfully',
        });
      }
    }

    return result;
  }

  private async readScriptExitCode(
    client: TrialClient,
    obs: Observation,
    waitSuccess: boolean
  ): Promise<number> {
    try {
      const response = await client.sandbox.readFile({
        path: JobExecutor.jobExitPath(client.trial.config),
      });
      const parsed = Number.parseInt(response.content.trim(), 10);
      if (Number.isInteger(parsed)) {
        return parsed;
      }
    } catch {
      // Fall back to the nohup protocol status for older or partially failed jobs.
    }
    return obs.exitCode ?? (waitSuccess ? 0 : 1);
  }

  /**
   * Build session environment — merge OSS_* vars from process with config.env.
   * Config values take precedence over process env.
   */
  static buildSessionEnv(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: Record<string, any>
  ): Record<string, string> | null {
    const ossEnv: Record<string, string> = {};
    for (const [k, v] of Object.entries(process.env)) {
      if (k.startsWith('OSS') && v !== undefined) {
        ossEnv[k] = v;
      }
    }

    const env = (config as Record<string, unknown>)['environment'] as Record<string, unknown>;
    const configEnv = (env?.['env'] ?? {}) as Record<string, string>;

    const merged = { ...ossEnv, ...configEnv };

    if (Object.keys(merged).length === 0) return null;
    return merged;
  }
}

File: rock/sdk/job/executor.py
"""JobExecutor drives Trials produced by an Operator."""

from __future__ import annotations

import asyncio
import os
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import TYPE_CHECKING

from rock.actions import CreateBashSessionRequest
from rock.logger import init_logger
from rock.sdk.job.operator import Operator
from rock.sdk.job.result import ExceptionInfo, TrialResult
from rock.sdk.sandbox.client import Sandbox

if TYPE_CHECKING:
    from rock.sdk.job.meta import JobMeta
    from rock.sdk.job.planner import PlannedJob
    from rock.sdk.job.config import JobConfig
    from rock.sdk.job.trial.abstract import AbstractTrial

logger = init_logger(__name__)


@dataclass
class TrialClient:
    sandbox: Sandbox
    session: str
    pid: int
    trial: AbstractTrial


@dataclass
class JobClient:
    trials: list[TrialClient]


TrialDoneCallback = Callable[[TrialClient, TrialResult | list[TrialResult], int], None]
TrialStartedCallback = Callable[[TrialClient, int], None]


class JobExecutor:
    """Execution engine for job trials.

    ``submit``/``wait`` preserve the historical single-job behavior.
    ``run_job`` is the atomic primitive used by CLI run orchestration.
    """

    def __init__(self, max_concurrent: int | None = None):
        if max_concurrent is not None and max_concurrent <= 0:
            raise ValueError("max_concurrent must be positive or None")
        self._max_concurrent = max_concurrent

    async def run(self, operator: Operator, config: JobConfig) -> list[TrialResult | list[TrialResult]]:
        job_client = await self.submit(operator, config)
        return await self.wait(job_client)

    async def run_job(
        self,
        job: PlannedJob,
        *,
        callbacks=None,
    ) -> TrialResult | list[TrialResult]:
        """Run a single planned job as an SDK atomic capability."""
        client = None
        try:
            client = await self._do_submit(job.trial)
            if callbacks and hasattr(callbacks, "on_started"):
                callbacks.on_started(client)
            result = await self._do_wait(client)
            if callbacks and hasattr(callbacks, "on_done"):
                callbacks.on_done(client, result)
            return result
        except Exception as exc:
            logger.error("Job lifecycle failed: %s", exc, exc_info=True)
            result = self._failure_result_for_trial(job.trial, exc)
            if client is not None and callbacks and hasattr(callbacks, "on_done"):
                callbacks.on_done(client, result)
            return result

    async def wait_existing_job(
        self,
        planned_job: PlannedJob,
        job_meta: JobMeta,
    ) -> TrialResult | list[TrialResult]:
        """Reconnect to an existing sandbox/process and collect the result.

        The CLI decides whether a job is recoverable; this method only performs
        the atomic wait/collect operation from persisted metadata.
        """
        if not job_meta.sandbox_id or not job_meta.session or job_meta.pid is None:
            raise ValueError("job_meta is missing sandbox_id/session/pid")
        sandbox = Sandbox(planned_job.config.environment)
        sandbox._sandbox_id = job_meta.sandbox_id
        client = TrialClient(
            sandbox=sandbox,
            session=job_meta.session,
            pid=job_meta.pid,
            trial=planned_job.trial,
        )
        return await self._do_wait(client)

    def prepare(self, operator: Operator, config: JobConfig) -> list[AbstractTrial]:
        return operator.apply(config)

    async def run_trials(
        self,
        trials: Sequence[AbstractTrial],
        on_trial_started: TrialStartedCallback | None = None,
        on_trial_done: TrialDoneCallback | None = None,
    ) -> list[TrialResult | list[TrialResult]]:
        """Run caller-provided trials with optional lifecycle concurrency control."""
        if not trials:
            return []

        semaphore = asyncio.Semaphore(self._max_concurrent) if self._max_concurrent is not None else None

        async def run_one(index: int, trial: AbstractTrial):
            if semaphore is None:
                return await self._run_trial_lifecycle(trial, index, on_trial_started, on_trial_done)
            async with semaphore:
                return await self._run_trial_lifecycle(trial, index, on_trial_started, on_trial_done)

        return list(await asyncio.gather(*[run_one(i, trial) for i, trial in enumerate(trials)]))

    async def submit(self, operator: Operator, config: JobConfig) -> JobClient:
        trial_list = self.prepare(operator, config)
        if not trial_list:
            return JobClient(trials=[])
        trial_clients = await asyncio.gather(*[self._do_submit(t) for t in trial_list])
        return JobClient(trials=list(trial_clients))

    async def wait(
        self,
        job_client: JobClient,
        on_trial_done: TrialDoneCallback | None = None,
    ) -> list[TrialResult | list[TrialResult]]:
        if not job_client.trials:
            return []
        if on_trial_done is None:
            return list(await asyncio.gather(*[self._do_wait(tc) for tc in job_client.trials]))

        async def wait_one(index: int, trial_client: TrialClient):
            result = await self._do_wait(trial_client)
            on_trial_done(trial_client, result, index)
            return result

        return list(await asyncio.gather(*[wait_one(i, tc) for i, tc in enumerate(job_client.trials)]))

    async def _run_trial_lifecycle(
        self,
        trial: AbstractTrial,
        index: int,
        on_trial_started: TrialStartedCallback | None,
        on_trial_done: TrialDoneCallback | None,
    ) -> TrialResult | list[TrialResult]:
        try:
            client = await self._do_submit(trial)
            if on_trial_started:
                on_trial_started(client, index)
            result = await self._do_wait(client)
            if on_trial_done:
                on_trial_done(client, result, index)
            return result
        except Exception as exc:
            logger.error("Trial lifecycle failed: %s", exc, exc_info=True)
            return self._failure_result_for_trial(trial, exc)

    @staticmethod
    def _failure_result_for_trial(trial: AbstractTrial, exc: Exception) -> TrialResult:
        config = getattr(trial, "_config", None)
        labels = getattr(config, "labels", {}) or {}
        task_name = labels.get("rock_task_id") or getattr(config, "job_name", "") or ""
        return TrialResult(
            task_name=task_name,
            exception_info=ExceptionInfo(
                exception_type=type(exc).__name__,
                exception_message=str(exc),
            ),
        )

    @staticmethod
    def _job_tmp_prefix(config: JobConfig) -> str:
        from rock.sdk.bench.constants import USER_DEFINED_LOGS

        return f"{USER_DEFINED_LOGS}/rock_job_{config.job_name or 'default'}"

    async def _do_submit(self, trial: AbstractTrial) -> TrialClient:
        config = trial._config
        sandbox = Sandbox(config.environment)
        await sandbox.start()
        logger.info("Sandbox started: sandbox_id=%s, job_name=%s", sandbox.sandbox_id, config.job_name)

        await trial.on_sandbox_ready(sandbox)
        await trial.setup(sandbox)

        session = f"rock-job-{config.job_name or 'default'}"
        env = self._build_session_env(config)
        await sandbox.create_session(CreateBashSessionRequest(session=session, env_enable=True, env=env))

        script_content = trial.build()
        prefix = self._job_tmp_prefix(config)
        script_path = f"{prefix}.sh"
        await sandbox.write_file_by_path(script_content, script_path)

        tmp_file = f"{prefix}.out"
        pid, error = await sandbox.start_nohup_process(
            cmd=f"bash {script_path}",
            tmp_file=tmp_file,
            session=session,
        )
        if error is not None:
            raise RuntimeError(f"Failed to start trial: {error.output}")

        logger.info("Trial started: pid=%s, job_name=%s", pid, config.job_name)
        return TrialClient(sandbox=sandbox, session=session, pid=pid, trial=trial)

    async def _do_wait(self, client: TrialClient) -> TrialResult | list[TrialResult]:
        config = client.trial._config
        success, message = await client.sandbox.wait_for_process_completion(
            pid=client.pid,
            session=client.session,
            wait_timeout=config.timeout,
            wait_interval=30,
        )
        obs = await client.sandbox.handle_nohup_output(
            tmp_file=f"{self._job_tmp_prefix(config)}.out",
            session=client.session,
            success=success,
            message=message,
            ignore_output=False,
            response_limited_bytes_in_nohup=None,
        )
        exit_code = obs.exit_code if obs.exit_code is not None else 1
        if obs.output:
            logger.info("Trial output (job=%s):\n%s", config.job_name, obs.output)
        result = await client.trial.collect(client.sandbox, obs.output or "", exit_code)
        iter_results = result if isinstance(result, list) else [result]
        for trial_result in iter_results:
            if not trial_result.raw_output:
                trial_result.raw_output = obs.output or ""
            if trial_result.exit_code == 0 and exit_code != 0:
                trial_result.exit_code = exit_code
        if not success:
            fail_info = ExceptionInfo(
                exception_type="ProcessTimeout",
                exception_message=message or "process did not complete successfully",
            )
            for trial_result in iter_results:
                if trial_result.exception_info is None:
                    trial_result.exception_info = fail_info
        return result

    @staticmethod
    def _build_session_env(config: JobConfig) -> dict[str, str] | None:
        oss_env = {k: v for k, v in os.environ.items() if k.startswith("OSS")}
        merged = {**oss_env, **config.environment.env}
        return merged or None
